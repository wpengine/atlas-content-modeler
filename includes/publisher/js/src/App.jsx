/* global atlasContentModelerFormEditingExperience */
import React, { useEffect, useState } from "react";
import Fields from "./components/Fields";
import { __ } from "@wordpress/i18n";
import { sendEvent } from "acm-analytics";
import TrashPostModal from "./components/TrashPostModal";

export default function App({ model, mode }) {
	const isEditMode = mode === "edit";
	const [trashPostModalIsOpen, setTrashPostModalIsOpen] = useState(false);

	useEffect(() => {
		const queryParams = new URLSearchParams(window.location.search);
		const newPost = queryParams.get("acm-post-published");
		if (newPost) {
			sendEvent({
				category: "post",
				action: "Post Published",
			});
		}
	}, []);

	return (
		<div className="app classic-form" style={{ marginTop: "20px" }}>
			<div className="flex-parent">
				<div>
					<h3 className="main-title">
						{isEditMode ? (
							<span>{__("Edit", "atlas-content-modeler")} </span>
						) : (
							<span>{__("Add", "atlas-content-modeler")} </span>
						)}
						{model.singular}
					</h3>
				</div>

				{isEditMode && (
					<div
						style={{ marginLeft: "20px" }}
						className="flex-align-v"
					>
						<a
							className="page-title-action"
							href={
								"/wp-admin/post-new.php?post_type=" + model.slug
							}
						>
							Add New
						</a>
					</div>
				)}
			</div>
			<div className="d-flex flex-column">
				<Fields model={model} />
			</div>
			<TrashPostModal
				isOpen={trashPostModalIsOpen}
				setIsOpen={setTrashPostModalIsOpen}
			/>
		</div>
	);
}
