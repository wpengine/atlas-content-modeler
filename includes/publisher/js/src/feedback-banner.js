const { apiFetch } = window.wp;

apiFetch({
	path: "/wpe/feedback-meta",
	method: "POST",
	_wpnonce: wpApiSettings.nonce,
})
	.then((res) => {
		console.log("response", res);
		if (res.success) {
			// dispatch({ type: "addModel", data: res.model });
			// history.push(
			// 	"/wp-admin/admin.php?page=atlas-content-modeler&view=edit-model&id=" +
			// 	data.slug
			// );
			//
			// // Insert the sidebar menu item below the Comments item, to avoid doing a full page refresh.
			// insertSidebarMenuItem(res.model);
			//
			// window.scrollTo(0, 0);
			// showSuccess(
			// 	`The “${res.model.plural}” model was created. Now add your first field.`
			// );
		}
	})
	.catch((err) => {
		// if (err.code === "atlas_content_modeler_already_exists") {
		// 	setError("slug", {
		// 		type: "idExists",
		// 		message: err.message,
		// 	});
		// }
	});
