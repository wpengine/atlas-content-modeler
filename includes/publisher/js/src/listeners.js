/**
 * Event listeners to adapt default WordPress publish page behavior.
 */

/**
 * Adds a 'submitted' class when the post form is submitted.
 *
 * Allows styling of :invalid input fields only when the form was submitted.
 * Prevents an issue where error messages appear for required fields when the
 * form is first loaded.
 */
export function postSubmittedListener() {
	const form = document.querySelector("form#post");
	const publishButton = document.querySelector("input#publish");
	const addSubmitted = () => form.classList.add("submitted");

	publishButton.addEventListener("click", addSubmitted);
	form.addEventListener("submit", addSubmitted);
}

/**
 * Calls the callback if the “Move To Trash” link is clicked.
 */
export function moveToTrashListener() {
	const moveToTrashLink = document.querySelector("#delete-action a");

	moveToTrashLink.addEventListener("click", (e) => {
		if (atlasContentModelerFormEditingExperience?.postHasReferences) {
			e.preventDefault();

			if (confirm("Are you sure you want to delete this?")) {
				window.location = moveToTrashLink.getAttribute("href");
			}
		}
	});
}
