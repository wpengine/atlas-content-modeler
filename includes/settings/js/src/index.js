import React from 'react';
import ReactDOM from 'react-dom';
import '../../scss/index.scss';

function App() {
	return (
		<p>This is the App component.</p>
	);
}

ReactDOM.render(
	<App />,
	document.getElementById('root')
);
