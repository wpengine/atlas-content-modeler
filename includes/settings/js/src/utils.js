import { useLocation } from "react-router-dom";
import { getFieldOrder, sanitizeFields } from "./queries";
import { toGraphQLType } from "./formats";
import { __ } from "@wordpress/i18n";

/**
 * Parses query string and returns value.
 *
 * @credit https://reactrouter.com/web/example/query-parameters
 * @returns {URLSearchParams}
 */
export function useLocationSearch() {
	return new URLSearchParams(useLocation().search);
}

/**
 * Inserts the content model list item in the wp-admin sidebar menu.
 *
 * @param {Object} model - The content model to be added to the sidebar.
 */
export function insertSidebarMenuItem(model) {
	const postMenuItems = document.querySelectorAll("[id^='menu-posts-']");
	let menuItem =
		postMenuItems.length > 0
			? postMenuItems[postMenuItems.length - 1]
			: document.getElementById("menu-comments");
	const markup = generateSidebarMenuItem(model);
	menuItem.insertAdjacentHTML("afterend", markup);
}

/**
 * Inserts the content model list item in the wp-admin top admin menu.
 *
 * @param {Object} model - The content model to be added to the admin menu.
 */
export function insertAdminMenuItem(model) {
	const postAdminMenuItem = document.querySelector(
		"[id='wp-admin-bar-new-content-default']"
	);
	const markup = generateAdminBarMenuItem(model);
	postAdminMenuItem.lastChild.insertAdjacentHTML("afterend", markup);
}

/**
 * Removes the post type from the wp-admin admin menu.
 *
 * @param {String} slug - The post type slug of the item to be removed.
 */
export function removeAdminMenuItem(slug) {
	const menuItem = document.querySelector(`[id="wp-admin-bar-new-${slug}"]`);
	if (menuItem) {
		menuItem.remove();
	}
}

/**
 * Removes the post type from the wp-admin sidebar menu.
 *
 * @param {String} slug - The post type slug of the item to be removed.
 */
export function removeSidebarMenuItem(slug) {
	const menuItem = document.querySelector(`[id="menu-posts-${slug}"]`);
	if (menuItem) {
		menuItem.remove();
	}
}

/**
 * Updates the sidebar post type menu item if the icon or label changed.
 *
 * @param {Object} model - The original content model information.
 * @param {Object} data - The updated content model data posted from the edit form.
 */
export function updateSidebarMenuItem(model, data) {
	const { model_icon = "dashicons-admin-post", plural } = data;

	// Update the post type menu label.
	if (model.plural !== plural) {
		document.querySelector(
			`li#menu-posts-${model.slug} .wp-menu-name`
		).innerHTML = plural;
	}

	// Update the post type menu icon.
	if (model.model_icon !== model_icon) {
		const menuIcon = document.querySelector(
			`li#menu-posts-${model.slug} .wp-menu-image`
		);

		menuIcon.classList.remove(model.model_icon);
		menuIcon.classList.add(model_icon);
	}
}

/**
 * Updates the admin post type menu item if the label changed.
 *
 * @param {Object} model - The original content model information.
 * @param {Object} data - The updated content model data posted from the edit form.
 */
export function updateAdminMenuItem(model, data) {
	const { singular } = data;

	// Update the post type menu label.
	if (model.singular !== singular) {
		document.querySelector(
			`li#wp-admin-bar-new-${model.slug} .ab-item`
		).innerHTML = singular;
	}
}

/**
 * Generates the HTML for the admin bar content model menu item.
 * @param {*} model
 */
export function generateAdminBarMenuItem(model) {
	let { slug, singular } = model;
	return `<li id="wp-admin-bar-new-${slug}"><a class="ab-item" href="post-new.php?post_type=${slug}">${singular}</a></li>`;
}

/**
 * Generates the HTML for the content model menu item.
 *
 * @param {Object} model - The content model.
 * @returns {string} - HTML list item markup for the specified content model.
 */
export function generateSidebarMenuItem(model) {
	let { slug, plural, model_icon } = model;
	if (!model_icon) {
		model_icon = "dashicons-admin-post";
	}

	return `<li class="wp-has-submenu wp-not-current-submenu menu-top menu-icon-${slug}" id="menu-posts-${slug}">
				<a href="edit.php?post_type=${slug}" class="wp-has-submenu wp-not-current-submenu menu-top menu-icon-${slug}" aria-haspopup="true">
					<div class="wp-menu-arrow">
						<div></div>
					</div>
					<div class="wp-menu-image dashicons-before ${model_icon}" aria-hidden="true"><br></div>
					<div class="wp-menu-name">${plural}</div>
				</a>
				<ul class="wp-submenu wp-submenu-wrap">
					<li class="wp-submenu-head" aria-hidden="true">${plural}</li>
					<li class="wp-first-item">
						<a href="edit.php?post_type=${slug}" class="wp-first-item">All ${plural}</a>
					</li>
					<li>
						<a href="post-new.php?post_type=${slug}">${__(
		"Add New",
		"atlas-content-modeler"
	)}</a>
					</li>
				</ul>
			</li>`;
}

/**
 * Closes the options dropdown if dropdown links are not in focus.
 *
 * @param {function} setDropdownOpen Call to toggle dropdown state.
 * @param {object} timer A ref to assign the timeout to. Allows cancellation when the calling component unmounts.
 */
export const maybeCloseDropdown = (setDropdownOpen, timer) => {
	timer.current = setTimeout(() => {
		const dropDownLinkIsInFocus = document?.activeElement?.parentElement.className.startsWith(
			"dropdown-content"
		);
		if (!dropDownLinkIsInFocus) {
			setDropdownOpen(false);
		}
	}, 100);
};

/**
 * Generates a link to open WPGraphQL's GraphiQL query editor in WP admin.
 *
 * Prefills the GraphiQL query with a request for the first 10 posts of the
 * `modelData` post type, including all fields in the saved field order.
 *
 * @param {object} modelData The full model data to generate a query from.
 * @return {string} The GraphiQL URL with query param prefilled.
 */
export const getGraphiQLLink = (modelData) => {
	const graphQLType = toGraphQLType(modelData.singular, true);
	const fragmentName = `${graphQLType}Fields`;
	const rootToTypeConnectionFieldName =
		modelData.singular !== modelData.plural
			? toGraphQLType(modelData.plural)
			: "all" + graphQLType;

	const fields = sanitizeFields(modelData?.fields);
	const fieldSlugs = getFieldOrder(fields).map((id) => {
		if (fields[id]?.type === "media") {
			return `
${fields[id]?.slug} {
  mediaItemId
  mediaItemUrl
  altText
  caption
  description
  mediaDetails {
    height
    width
    sizes {
      file
      fileSize
      height
      mimeType
      name
      sourceUrl
      width
    }
  }
}
`;
		}

		return fields[id]?.slug;
	});

	if (fieldSlugs.length === 0) {
		fieldSlugs.push("databaseId");
	}

	const query = `
{
  ${rootToTypeConnectionFieldName}(first: 10) {
    nodes {
      ...${fragmentName}
    }
  }
}

fragment ${fragmentName} on ${graphQLType} {
  ${fieldSlugs.join("\n  ")}
}
`;

	return `/wp-admin/admin.php?page=graphiql-ide&explorerIsOpen=true&query=${encodeURIComponent(
		query
	)}`;
};
