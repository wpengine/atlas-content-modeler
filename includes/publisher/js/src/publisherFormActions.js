import { getFieldOrder } from "../../../settings/js/src/queries";
const { postType, models } = atlasContentModelerFormEditingExperience;

/**
 * Adds a "submitted" class when the publish form is submitted.
 *
 * Allows styling of :invalid input fields only when the form was
 * submitted. Prevents an issue where error messages appear for
 * required fields when the form is first loaded.
 */
export const addClassOnFormSubmit = () => {
	const form = document.querySelector("form#post");
	const publishButton = document.querySelector("input#publish");
	const addSubmittedClass = () => form.classList.add("submitted");

	publishButton.addEventListener("click", addSubmittedClass);
	form.addEventListener("submit", addSubmittedClass);
};

/**
 * Puts focus on the first field.
 *
 * Allows the user to start interacting with the form immediately,
 * the same behavior as WordPress core post types.
 *
 * Acts when creating new posts, not when editing existing ones.
 */
export const focusFirstField = () => {
	if (!document.body.classList.contains("post-new-php")) {
		return;
	}

	const fieldOrder = getFieldOrder(models[postType]?.fields);
	const firstFieldId = fieldOrder[0] ?? false;
	const firstField = firstFieldId
		? models[postType]?.fields[firstFieldId]
		: false;

	if (!firstField) {
		return;
	}

	const { type, inputType } = firstField;

	if (type === "richtext") {
		return; // Rich Text field focus is handled in the useWPEditor hook.
	}

	let selector = "input";

	if (type === "text" && inputType === "multi") {
		selector = "textarea";
	}

	if (type === "media" || type === "relationship") {
		selector = "button";
	}

	const field = document.querySelector(
		`div[data-first-field=true] ${selector}`
	);

	if (field) {
		field.focus();
	}
};
