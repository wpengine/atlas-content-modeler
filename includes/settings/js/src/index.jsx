import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import "../../scss/index.scss";

ReactDOM.render(<App />, document.getElementById("root"));
