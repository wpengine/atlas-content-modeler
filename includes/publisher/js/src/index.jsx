/* global atlasContentModelerFormEditingExperience */
import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import "./../../scss/index.scss";
import "../../../settings/scss/index.scss";
import { addClassOnFormSubmit, focusFirstField } from "./publisherFormActions";

const { models, postType } = atlasContentModelerFormEditingExperience;
const container = document.getElementById("atlas-content-modeler-fields-app");

if (container && models.hasOwnProperty(postType)) {
	const model = models[postType];
	const urlParams = new URLSearchParams(window.location.search);

	ReactDOM.render(
		<App model={model} mode={urlParams.get("action")} />,
		container
	);

	addClassOnFormSubmit();
	focusFirstField();
}
