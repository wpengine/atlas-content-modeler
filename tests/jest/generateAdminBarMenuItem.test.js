import { generateAdminBarMenuItem } from "../../includes/settings/js/src/utils";

describe("generateAdminBarMenuItem tests", () => {
	const mock = {
		slug: "cows",
		singular: "Cow",
	};

	it("Renders a matching snapshot", () => {
		const markup = generateAdminBarMenuItem(mock);
		expect(markup).toMatchSnapshot();
	});
});
