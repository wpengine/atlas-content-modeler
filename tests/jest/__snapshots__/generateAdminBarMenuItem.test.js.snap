// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAdminBarMenuItem tests Renders a matching snapshot 1`] = `"<li id=\\"wp-admin-bar-new-cows\\"><a class=\\"ab-item\\" href=\\"post-new.php?post_type=cows\\">Cow</a></li>"`;
